stdlib/condition.mli

```diff
@@ -13,41 +13,112 @@
+   nonempty. A thread that inserts an element into the queue signals
+   that the queue is nonempty. A condition variable is used for this
+   purpose. This communication channel conveys the information that
+   the property "the queue is nonempty" is true, or (more accurately)
+   may be true.
+
+   In short, a condition variable [c] is used to convey the information
+   that a certain property [P] about a shared data structure [D],
+   protected by a mutex [m], may be true.
+
+   With a condition variable [c], exactly one mutex [m] is associated.
+   This association is implicit: the mutex [m] is not explicitly passed
+   as an argument to {!create}. It is up to the programmer to know, for
+   each condition variable [c], which is the associated mutex [m].
+
+   With a mutex [m], several condition variables can be associated.
+   For instance, in a bounded queue, one condition variable could
+   be used to indicate that the queue is nonempty, and another condition
+   variable could be used to indicate that the queue is not full.
+
+   With a condition variable [c], exactly one logical property [P]
+   should be associated. Examples of such properties
+   include "the queue is nonempty" and "the queue is not full".
+   It is up to the programmer to keep track, for each condition
+   variable, of the corresponding property [P].
+   A signal should be sent on the condition variable [c]
+   only when the property [P] is definitely true.
+   When a signal is received, however, one cannot assume that [P] is
+   definitely true; one must explicitly test whether [P] is true.
+   The reason is that,
+   between the moment when the signal is sent
+   and the moment when a waiting thread receives the signal
+   and is scheduled,
+   the property [P] may be falsified by some other thread
+   that is able to acquire the mutex [m] and alter the data structure [D].
+
+   Assuming that [D] is a shared data structure
+   protected by the mutex [m],
+   and assuming that [c] is a condition variable
+   associated with the mutex [m]
+   and with the property [P],
+   a typical usage scenario is as follows:
    {[
      Mutex.lock m;
-     while (* some predicate P over D is not satisfied *) do
+     while (* the property P over D is not satisfied *) do
        Condition.wait c m
      done;
      (* Modify D *)
-     if (* the predicate P over D is now satisfied *) then Condition.signal c;
+     if (* the property P over D is now satisfied *) then Condition.signal c;
      Mutex.unlock m
    ]}
+
+   {!signal} and {!broadcast} should be called only when the mutex [m]
+   associated with the condition variable [c] is locked.
+   Otherwise, there would be a risk that some other thread is currently
+   inside a critical section, has determined that the property [P]
+   associated with [c] is false, but has not yet called [wait]. In such a
+   case, the signal would be lost, leading to a potential deadlock.
```

Comment:
I am not so sure about this, it is common practice to call {!signal} and {!broadcast} outside of critical sections.

As far as I understand it, the implementation can take care of the involved mutex as it is passed as an argument ro wait. Also notice that, in POSIX threads, it is an error to invoke {! wait} on the same condition variable with different mutexes. See the error section  [here](https://linux.die.net/man/3/pthread_cond_wait).

Reply:
POSIX threads definitely support using signal and broadcast outside of the critical section, and it's not clear to me how this can lead to a deadlock.  However, common wisdom is that signaling inside the critical section leads to better scheduling and more fair executions overall.

Reply:
If a thread A uses `signal` or `broadcast` without holding the lock, then one runs the risk that a thread B might be inside a critical section, immediately after determining that the desired property is false, and before calling `wait`. Because no thread is currently waiting, the signal is lost. Then, thread B goes to sleep, even though the desired condition is in fact true. This can lead to a situation such as a consumer waiting indefinitely to extract an item out of a queue, when the queue is in fact nonempty: this is what I call a deadlock.

Furthermore, it seems dubious to call `signal` or `broadcast` outside a critical section, since one cannot know in that case whether the desired property is true.

Finally, I note that in Java, `notify` and `notifyAll` can be called only when the lock is held.

This is why I suggest requiring that the lock be held when `signal` or `broadcast` is called. But, if you insist, I can change this to a recommendation instead of a requirement.

Reply:
> If a thread A uses `signal` or `broadcast` without holding the lock, then one runs the risk that a thread B might be inside a critical section, immediately after determining that the desired property is false, and before calling `wait`. Because no thread is currently waiting, the signal is lost. Then, thread B goes to sleep, even though the desired condition is in fact true. This can lead to a situation such as a consumer waiting indefinitely to extract an item out of a queue, when the queue is in fact nonempty: this is what I call a deadlock.

You have to describe a complete scenario to convince us . For instance, I do not understand what you mean by "The signal is lost". The fact is that POSIX programmers agree that signaling from outside of critical section is possible, although not recommended. And again, one can signal when the property is false, even inside critical sections (see comment below).

Reply:
Actually, my scenario is complete already: thread A signals, the signal is lost, thread B waits indefinitely for a signal, even though the property desired by thread B is already true.

I agree that signalling outside a critical section does not compromise *safety*, but from the point of view of *liveness*, a signal sent outside of a critical section can be lost and therefore is just as good as no signal at all. In other words, when proving absence of deadlocks, a call to `signal` or `broadcast` outside of a critical section cannot contribute to the proof.

This is why I suggest to at least *strongly recommend* to use these functions inside a critical section.

Reply:
May I beg to differ: you assume some implementation of {! signal} that is not described. I am not sure that a POSIX compliant implementation could legitimately lose the signal here, although such implementation does not look that easy to design at first sight.

Reply:
Actually, this version of the [standard](https://pubs.opengroup.org/onlinepubs/7908799/xsh/pthread_cond_signal.html) says: *the `pthread_cond_signal()` and `pthread_cond_broadcast()` functions have no effect if there are no threads currently blocked on cond.* This suggests that the signal is lost if no thread is currently waiting. (And it seems to be common wisdom that this problem exists; it is known as a *lost wakeup*.)

The same document also indicates that it is OK to use `signal` and `broadcast` without holding the mutex, yet it adds: *however, if predictable scheduling behaviour is required, then the mutex is locked by the thread calling `signal` or `broadcast`*. It is unclear what this means...

Reply:
Thanks for the "lost wakeup", I did not know that this situation has a name. Looking for it on the web I [found](https://docs.oracle.com/cd/E19120-01/open.solaris/816-5137/sync-30/index.html):

> This can occur only if the condition being tested is modified without holding the mutex lock associated with the condition. As long as the condition being tested is modified only while holding the associated mutex, pthread_cond_signal() and pthread_cond_broadcast() can be called regardless of whether they are holding the associated mutex.

For what it's worth, I would interpret that lost wakeup may occur in racy programs. As far as POSIX is concerned, anything can happen.  For OCaml, the situation may be different, as racy programs are allowed.

My current opinion  is that the question is a bit too complex to be answered in `condition.mli` comments, and that I'll continue to call {! signal} inside critical sections. I would tend to recommend avoiding the practice of signalling from inside critical section more for code clarity than anything else.

Reply:
I have thought about the sentence that you quote (*As long as the condition being tested is modified only while holding the associated mutex, [...]*). I believe it may be true, although I am unable to construct a clear argument. (I think it is time I read [this paper](https://people.cs.kuleuven.be/~bart.jacobs/esop18.pdf)...) Anyway, I will change the text to remove the requirement that the lock be held.